### DISCUSSION TITLE
关于 struct_pack struct_json 序列化的问题请教

### DISCUSSION THREAD

**Discussion by Weaverzhu** [alibaba/yalantinglibs#701] [Q&A] [answered]
我在为一个很大的 codebase 支持一些结构体的序列化工作，关于 struct_pack struct_json 序列化的问题想请假一下：

1. struct_pb 和 [struct_json ](https://alibaba.github.io/yalantinglibs/zh/struct_xxx/struct_xxx_introduction.html#json-%E5%BA%8F%E5%88%97%E5%8C%96-%E5%8F%8D%E5%BA%8F%E5%88%97%E5%8C%96) 是否支持像 [struct_pack ](https://alibaba.github.io/yalantinglibs/zh/struct_pack/struct_pack_intro.html#struct-pack%E7%AE%80%E4%BB%8B) 那样自动化的反射？我遇到的问题是：如果是一个很大的 struct 需要迁移到 struct_pb，为它编写反射是一件非常冗长的工作。
    但看起来也有点不切实际，感觉上 struct_pack 依靠一些假设，没有保留 key name 信息，但是 json 和 pb 中分别由 key string 和 field id；所以可以说 struct_json 一定需要用户来编写反射吗，或者最多编写 .h -> 编写 proto，让 codegen 来完成反射
3. struct_pack 的自动化反射是否有缺点？比如编译速度在 struct 字段数多大时，可能会对用户造成负担？（e.g. 单 cpp 超过 10s）
4. 如果我的 struct 不包括基本类型，请问反射是否提供一些绕过机制？比如
    ```cpp
    struct Foo {
      nlohmann::json params;
    };
    ```
    这个 struct 如何和 struct_json 方案融合？
5. struct_json 是否也有 [compatible 字段](https://alibaba.github.io/yalantinglibs/zh/struct_pack/struct_pack_intro.html#%E5%90%91%E5%89%8D-%E5%90%91%E5%90%8E%E5%85%BC%E5%AE%B9%E6%80%A7) 这样的兼容性机制

**Comment by poor-circle** [comment, marked as answer]
1. struct_json不太可行，因为目前c++的反射机制不支持获取成员的名字，得等C++的官方的反射题案。但是struct_pb应该是可以支持的，field_id可以自动生成，后面会完善。
2. 实际上，我们在clang下做过编译时间分析，在O3时，编译器前端（包含你说的反射）占用的时间大约只有10%，大头在中后端（代码优化和生成）。字段非常多的情况下，struct_pack编译较慢的可能原因是，因为大量的内联和模板展开导致的代码膨胀。这种情况下，我们建议：
```cpp
//serialize.hpp
std::string my_serialize(const Foo &foo);
```
```cpp
//serialize.cpp
std::string my_serialize(const Foo &foo) {
  return struct_pack::serialize<std::string>(foo);
}
```
使用时调用my_serialize而非struct_pack::serialize，这样你只需要编译一次serialize.cpp即可。当然代价是多了一次无法内联的函数调用。
3. struct_pack有相关机制，请见[介绍](https://alibaba.github.io/yalantinglibs/zh/struct_pack/struct_pack_intro.html#%E8%87%AA%E5%AE%9A%E4%B9%89%E7%B1%BB%E5%9E%8B%E7%9A%84%E5%BA%8F%E5%88%97%E5%8C%96)。struct_json目前恐怕是不支持的。
4. 无。但我理解json格式天然就是向前向后兼容的。多余的字段应该会被自动忽略，缺失的字段应该会被设置为默认值。

**Reply by Weaverzhu** [reply]
> 1. struct_json不太可行，因为目前c++的反射机制不支持获取成员的名字，得等C++的官方的反射题案。但是struct_pb应该是可以支持的，field_id可以自动生成，后面会完善。

今天刚看到了一个 lib https://github.com/getml/reflect-cpp 感觉能在编译器拿到 field name，这个东西能 apply 到 struct_json 吗

**Reply by Weaverzhu** [reply]
https://veselink1.github.io/blog/cpp/metaprogramming/2019/07/13/refl-cpp-deep-dive.html 甚至好像找到了一篇教程

**Reply by qicosmos** [reply]
应该是可以的，晚点试一下。

**Reply by poor-circle** [reply]
> > 1. struct_json不太可行，因为目前c++的反射机制不支持获取成员的名字，得等C++的官方的反射题案。但是struct_pb应该是可以支持的，field_id可以自动生成，后面会完善。
> 
> 今天刚看到了一个 lib https://github.com/getml/reflect-cpp 感觉能在编译器拿到 field name，这个东西能 apply 到 struct_json 吗

非常感谢，稍后会重构，让他支持不需要宏

**Comment by qicosmos** [comment]
@Weaverzhu 
struct_json 因为需要用到字段名，在C++编译期反射进入标准之前还需要宏。
另外，生成反射宏可以使用iguana提供的脚本去生成，不需要手写：https://github.com/qicosmos/iguana?tab=readme-ov-file#scripts
